crates/sozo/ops/src/migrate/mod.rs

```diff
@@ -103,6 +108,83 @@ where
         })
     }
 
+    /// Upload resources metadata to IPFS and update the ResourceMetadata Dojo model.
+    ///
+    /// # Arguments
+    ///
+    /// # Returns
+    ///
+    pub async fn upload_metadata(&self, ui: &mut MigrationUi) -> anyhow::Result<()> {
+        ui.update_text("Uploading metadata...");
+
+        let mut invoker = Invoker::new(&self.world.account, self.txn_config);
+
+        // world
+        let stored_metadata = self.world.metadata(&Felt::ZERO).call().await?;
+        let new_metadata = WorldMetadata::from(self.diff.profile_config.world.clone());
+
+        let res = new_metadata.upload_if_changed(stored_metadata.metadata_hash).await?;
+
+        if let Some((new_uri, new_hash)) = res {
+            invoker.add_call(self.world.set_metadata_getcall(&ResourceMetadata {
+                resource_id: Felt::ZERO,
+                metadata_uri: ByteArray::from_string(&new_uri)?,
+                metadata_hash: new_hash,
+            }));
+        }
+
+        // contracts
+        if let Some(configs) = &self.diff.profile_config.contracts {
+            let calls = self.upload_metadata_from_resource_config(configs).await?;
+            invoker.extend_calls(calls);
+        }
+
+        // models
+        if let Some(configs) = &self.diff.profile_config.models {
+            let calls = self.upload_metadata_from_resource_config(configs).await?;
+            invoker.extend_calls(calls);
+        }
+
+        if self.do_multicall() {
+            ui.update_text_boxed(format!("Uploading {} metadata...", invoker.calls.len()));
+            invoker.multicall().await.or_else(|e| Err(anyhow!(e.to_string())))?;
+        } else {
+            ui.update_text_boxed(format!(
+                "Uploading {} metadata (sequentially)...",
+                invoker.calls.len()
+            ));
+            invoker.invoke_all_sequentially().await.or_else(|e| Err(anyhow!(e.to_string())))?;
+        }
+
+        Ok(())
+    }
+
+    async fn upload_metadata_from_resource_config(
+        &self,
+        config: &Vec<ResourceConfig>,
+    ) -> anyhow::Result<Vec<Call>> {
+        let mut calls = vec![];
+
+        for item in config {
+            let selector = dojo_types::naming::compute_selector_from_tag_or_name(&item.tag);
+
+            let stored_metadata = self.world.metadata(&selector).call().await?;
```

Comment:
Since the migration is fetching the world already, and we have an event for metadata, we may avoid fetching this if we use the actual event.

Using `ResourceRemote`, we can add a metadata field. And we update it when we fetch the events rebuilding the remote world: https://github.com/dojoengine/dojo/blob/51d80cf369a529f7ecfe3c02e9f42d96cea6e8a0/crates/dojo/world/src/remote/events_to_remote.rs#L123

Doing so, we could avoid this fetch, and only compare with the remote world state?